File: ya/memory.py
from __future__ import annotations

from dataclasses import asdict, dataclass
from datetime import datetime, timezone
from pathlib import Path
import re
import unicodedata
from uuid import uuid4
import json

from .config import data_home


MAX_MEMORY_CARDS = 100
MAX_RELEVANT_MEMORY_CARDS = 3
MIN_RELEVANCE_SCORE = 3
ACTIVE_MEMORY_STATUSES = {"candidate", "approved"}
DEFAULT_PRUNE_STATUSES = {"rejected", "revoked"}
ENGLISH_WORD = re.compile(r"[a-z0-9][a-z0-9_-]*")
HAN_TEXT = re.compile(r"[\u4e00-\u9fff]+")
ENGLISH_STOP_WORDS = {
    "a", "an", "and", "are", "as", "at", "be", "by", "for", "from", "how", "in", "is",
    "it", "of", "on", "or", "the", "this", "that", "to", "what", "when", "where", "with",
}
CHINESE_STOP_PHRASES = {
    "什么", "如何", "为什", "什么是", "怎么", "可以", "请问", "一个", "这个", "那个",
    "我们", "你们", "他们", "关于", "以及", "进行", "一下", "是否", "需要",
}


class DuplicateMemoryError(ValueError):
    def __init__(self, card: "MemoryCard") -> None:
        self.card = card
        super().__init__(f"Matching memory card already exists: {card.id}")


class MemoryLimitError(ValueError):
    pass


@dataclass
class MemoryCard:
    id: str
    kind: str
    text: str
    evidence: str
    status: str
    created_at: str
    version: int = 1


@dataclass(frozen=True)
class MemoryMatch:
    card: MemoryCard
    score: int


def memory_path() -> Path:
    return data_home() / "memory.json"


def _load() -> list[MemoryCard]:
    path = memory_path()
    if not path.exists():
        return []
    with path.open(encoding="utf-8") as handle:
        return [MemoryCard(**item) for item in json.load(handle)]


def _save(cards: list[MemoryCard]) -> None:
    path = memory_path()
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("w", encoding="utf-8") as handle:
        json.dump([asdict(card) for card in cards], handle, ensure_ascii=False, indent=2)


def normalize_memory_text(text: str) -> str:
    normalized = unicodedata.normalize("NFKC", text).casefold()
    return re.sub(r"\s+", " ", normalized).strip()


def create_candidate(text: str, evidence: str, kind: str = "procedure") -> MemoryCard:
    if kind not in {"preference", "procedure", "knowledge"}:
        raise ValueError("memory kind must be preference, procedure, or knowledge.")
    cards = _load()
    normalized_text = normalize_memory_text(text)
    for existing in cards:
        if (
            existing.kind == kind
            and existing.status in ACTIVE_MEMORY_STATUSES
            and normalize_memory_text(existing.text) == normalized_text
        ):
            raise DuplicateMemoryError(existing)
    if len(cards) >= MAX_MEMORY_CARDS:
        raise MemoryLimitError(
            f"Memory card limit ({MAX_MEMORY_CARDS}) reached. Run: ya memory prune"
        )
    card = MemoryCard(
        id=uuid4().hex[:8],
        kind=kind,
        text=text.strip(),
        evidence=evidence.strip(),
        status="candidate",
        created_at=datetime.now(timezone.utc).isoformat(),
    )
    cards.append(card)
    _save(cards)
    return card


def list_cards(status: str | None = None) -> list[MemoryCard]:
    cards = _load()
    return [card for card in cards if card.status == status] if status else cards


def set_status(card_id: str, status: str) -> MemoryCard:
    if status not in {"approved", "rejected", "revoked"}:
        raise ValueError("invalid memory status")
    cards = _load()
    for card in cards:
        if card.id == card_id:
            card.status = status
            card.version += 1
            _save(cards)
            return card
    raise ValueError("memory card not found")


def cards_to_prune(include_candidates: bool = False) -> list[MemoryCard]:
    statuses = set(DEFAULT_PRUNE_STATUSES)
    if include_candidates:
        statuses.add("candidate")
    return [card for card in _load() if card.status in statuses]


def prune_cards(include_candidates: bool = False) -> list[MemoryCard]:
    statuses = set(DEFAULT_PRUNE_STATUSES)
    if include_candidates:
        statuses.add("candidate")
    cards = _load()
    removed = [card for card in cards if card.status in statuses]
    if removed:
        _save([card for card in cards if card.status not in statuses])
    return removed


def _english_words(text: str) -> set[str]:
    return {
        word
        for word in ENGLISH_WORD.findall(text)
        if len(word) >= 2 and word not in ENGLISH_STOP_WORDS
    }


def _english_phrases(text: str) -> set[str]:
    words = [word for word in ENGLISH_WORD.findall(text) if word not in ENGLISH_STOP_WORDS]
    return {" ".join(words[index : index + 2]) for index in range(len(words) - 1)}


def _han_ngrams(text: str, width: int) -> set[str]:
    grams: set[str] = set()
    for run in HAN_TEXT.findall(text):
        grams.update(run[index : index + width] for index in range(len(run) - width + 1))
    return {gram for gram in grams if gram not in CHINESE_STOP_PHRASES}


def _memory_score(task: str, card: MemoryCard) -> int:
    task_text = normalize_memory_text(task)
    card_text = normalize_memory_text(card.text)
    if not task_text or not card_text:
        return 0

    shared_words = _english_words(task_text) & _english_words(card_text)
    shared_bigrams = _han_ngrams(task_text, 2) & _han_ngrams(card_text, 2)
    shared_phrases = (
        (_english_phrases(task_text) & _english_phrases(card_text))
        or (_han_ngrams(task_text, 3) & _han_ngrams(card_text, 3))
    )
    exact_containment = (
        min(len(task_text), len(card_text)) >= 4
        and (task_text in card_text or card_text in task_text)
    )
    phrase_score = 5 if shared_phrases or exact_containment else 0
    return phrase_score + 3 * len(shared_words) + len(shared_bigrams)


def _created_timestamp(card: MemoryCard) -> float:
    try:
        return datetime.fromisoformat(card.created_at.replace("Z", "+00:00")).timestamp()
    except ValueError:
        return 0.0


def select_relevant_cards(task: str, limit: int = MAX_RELEVANT_MEMORY_CARDS) -> list[MemoryMatch]:
    """Return approved cards ranked by local, deterministic task relevance."""
    matches = [
        MemoryMatch(card, score)
        for card in list_cards("approved")
        if (score := _memory_score(task, card)) >= MIN_RELEVANCE_SCORE
    ]
    matches.sort(key=lambda match: (-match.score, -_created_timestamp(match.card), match.card.id))
    return matches[:limit]


def relevant_context(task: str, limit: int = MAX_RELEVANT_MEMORY_CARDS) -> str:
    matches = select_relevant_cards(task, limit)
    if not matches:
        return ""
    lines = [f"- [{match.card.kind}] {match.card.text}" for match in matches]
    return "\n".join(lines)

File: src/memory.ts
import { randomUUID } from "node:crypto";
import { existsSync, mkdirSync, readFileSync, renameSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import {
  memoryScore as nativeMemoryScore,
  normalizeMemoryText as nativeNormalizeMemoryText,
} from "ya-core";
import { dataHome } from "./config";

export const MAX_MEMORY_CARDS = 100;
export const MAX_RELEVANT_MEMORY_CARDS = 3;
export const MIN_RELEVANCE_SCORE = 3;

export type MemoryKind = "preference" | "procedure" | "knowledge";
export type MemoryStatus = "candidate" | "approved" | "rejected" | "revoked";

export interface MemoryCard {
  id: string;
  kind: MemoryKind;
  text: string;
  evidence: string;
  status: MemoryStatus;
  createdAt: string;
  version: number;
}

interface StoredMemoryCard extends Omit<MemoryCard, "createdAt"> {
  created_at: string;
}

export interface MemoryMatch {
  card: MemoryCard;
  score: number;
}

const ACTIVE_STATUSES = new Set<MemoryStatus>(["candidate", "approved"]);
const DEFAULT_PRUNE_STATUSES = new Set<MemoryStatus>(["rejected", "revoked"]);

export class DuplicateMemoryError extends Error {
  readonly card: MemoryCard;

  constructor(card: MemoryCard) {
    super(`Matching memory card already exists: ${card.id}`);
    this.name = "DuplicateMemoryError";
    this.card = card;
  }
}

export class MemoryLimitError extends Error {
  override readonly name = "MemoryLimitError";
}

export function memoryPath(): string {
  return join(dataHome(), "memory.json");
}

function fromStored(value: unknown): MemoryCard {
  if (!isRecord(value)) throw new Error("Invalid memory card data.");
  return {
    id: expectString(value.id, "id"),
    kind: expectString(value.kind, "kind") as MemoryKind,
    text: expectString(value.text, "text"),
    evidence: expectString(value.evidence, "evidence"),
    status: expectString(value.status, "status") as MemoryStatus,
    createdAt: expectString(value.created_at ?? value.createdAt, "created_at"),
    version: typeof value.version === "number" ? value.version : 1,
  };
}

function toStored(card: MemoryCard): StoredMemoryCard {
  const { createdAt, ...rest } = card;
  return { ...rest, created_at: createdAt };
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function expectString(value: unknown, field: string): string {
  if (typeof value !== "string") throw new Error(`Invalid memory card ${field}.`);
  return value;
}

export function loadMemoryCards(): MemoryCard[] {
  const path = memoryPath();
  if (!existsSync(path)) return [];
  const value = JSON.parse(readFileSync(path, "utf8")) as unknown;
  if (!Array.isArray(value)) throw new Error("Memory file must contain a JSON array.");
  return value.map(fromStored);
}

export function saveMemoryCards(cards: readonly MemoryCard[]): void {
  const path = memoryPath();
  mkdirSync(dirname(path), { recursive: true });
  const temporary = `${path}.${process.pid}.tmp`;
  writeFileSync(temporary, `${JSON.stringify(cards.map(toStored), null, 2)}\n`, "utf8");
  renameSync(temporary, path);
}

/** NFKC, case folding, `ß` to `ss`, and collapsed whitespace; implemented in Rust. */
export function normalizeMemoryText(text: string): string {
  return nativeNormalizeMemoryText(text);
}

export function createCandidate(text: string, evidence: string, kind: MemoryKind = "procedure"): MemoryCard {
  if (!(["preference", "procedure", "knowledge"] as const).includes(kind)) {
    throw new Error("memory kind must be preference, procedure, or knowledge.");
  }
  const cards = loadMemoryCards();
  const normalizedText = normalizeMemoryText(text);
  const duplicate = cards.find(
    (card) => card.kind === kind && ACTIVE_STATUSES.has(card.status) && normalizeMemoryText(card.text) === normalizedText,
  );
  if (duplicate) throw new DuplicateMemoryError(duplicate);
  if (cards.length >= MAX_MEMORY_CARDS) {
    throw new MemoryLimitError(`Memory card limit (${MAX_MEMORY_CARDS}) reached. Run: ya memory prune`);
  }
  const card: MemoryCard = {
    id: randomUUID().replaceAll("-", "").slice(0, 8),
    kind,
    text: text.trim(),
    evidence: evidence.trim(),
    status: "candidate",
    createdAt: new Date().toISOString(),
    version: 1,
  };
  cards.push(card);
  saveMemoryCards(cards);
  return card;
}

export function listCards(status?: MemoryStatus): MemoryCard[] {
  const cards = loadMemoryCards();
  return status ? cards.filter((card) => card.status === status) : cards;
}

export function setStatus(cardId: string, status: Exclude<MemoryStatus, "candidate">): MemoryCard {
  if (!(["approved", "rejected", "revoked"] as const).includes(status)) throw new Error("invalid memory status");
  const cards = loadMemoryCards();
  const card = cards.find((candidate) => candidate.id === cardId);
  if (!card) throw new Error("memory card not found");
  card.status = status;
  card.version += 1;
  saveMemoryCards(cards);
  return card;
}

export function cardsToPrune(includeCandidates = false): MemoryCard[] {
  return loadMemoryCards().filter(
    (card) => DEFAULT_PRUNE_STATUSES.has(card.status) || (includeCandidates && card.status === "candidate"),
  );
}

export function pruneCards(includeCandidates = false): MemoryCard[] {
  const cards = loadMemoryCards();
  const removed = cards.filter(
    (card) => DEFAULT_PRUNE_STATUSES.has(card.status) || (includeCandidates && card.status === "candidate"),
  );
  if (removed.length > 0) {
    const removedIds = new Set(removed.map((card) => card.id));
    saveMemoryCards(cards.filter((card) => !removedIds.has(card.id)));
  }
  return removed;
}

/**
 * Scores a task against a card: shared English words, shared Han bigrams, and a
 * phrase or containment bonus. Implemented in Rust.
 */
export function memoryScore(task: string, card: MemoryCard): number {
  return nativeMemoryScore(task, card.text);
}

export function selectRelevantCards(task: string, limit = MAX_RELEVANT_MEMORY_CARDS): MemoryMatch[] {
  return listCards("approved")
    .map((card) => ({ card, score: memoryScore(task, card) }))
    .filter((match) => match.score >= MIN_RELEVANCE_SCORE)
    .sort((left, right) => {
      if (left.score !== right.score) return right.score - left.score;
      const dateDifference = (Date.parse(right.card.createdAt) || 0) - (Date.parse(left.card.createdAt) || 0);
      return dateDifference || left.card.id.localeCompare(right.card.id);
    })
    .slice(0, limit);
}

export function relevantContext(task: string, limit = MAX_RELEVANT_MEMORY_CARDS): string {
  return selectRelevantCards(task, limit).map(({ card }) => `- [${card.kind}] ${card.text}`).join("\n");
}
